### PR TITLE
Add C# code for the chapter Heap and Graph

### PR DESCRIPTION
**Pull request by ZJKung**
If this PR is related to coding or code translation, please read the [contribution guideline](https://github.com/krahets/hello-algo/issues/15#issue-1464069591), fill out the checklist, and paste the console outputs to the PR.

-  [X] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [X] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [X] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

**Comment by ZJKung** [comment]
Here is the test console output,please check.

# Test Output

-----
```
hello_algo.chapter_graph.graph_adjacency_list.Test:
    Outcome: Passed
    Standard Output Messages:
    
初始化后，图为
邻接表 =
1: 3 5,
3: 1 2,
2: 3 5 4,
5: 2 1 4,
4: 2 5,
添加边 1-2 后，图为
邻接表 =
1: 3 5 2,
3: 1 2,
2: 3 5 4 1,
5: 2 1 4,
4: 2 5,
删除边 1-3 后，图为
邻接表 =
1: 5 2,
3: 2,
2: 3 5 4 1,
5: 2 1 4,
4: 2 5,
添加顶点 6 后，图为
邻接表 =
1: 5 2,
3: 2,
2: 3 5 4 1,
5: 2 1 4,
4: 2 5,
6: ,
删除顶点 3 后，图为
邻接表 =
1: 5 2,
2: 5 4 1,
5: 2 1 4,
4: 2 5,
6: ,
```

---

```
hello_algo.chapter_graph.graph_adjacency_matrix.Test:
    Outcome: Passed
    Standard Output Messages:
    
初始化后，图为
顶点列表 = 
1,3,2,5,4
邻接矩阵 =
[
  0,  1,  1,  1,  0
  1,  0,  1,  0,  0
  1,  1,  0,  1,  1
  1,  0,  1,  0,  1
  0,  0,  1,  1,  0
]
添加边 1-2 后，图为
顶点列表 = 
1,3,2,5,4
邻接矩阵 =
[
  0,  1,  1,  1,  0
  1,  0,  1,  0,  0
  1,  1,  0,  1,  1
  1,  0,  1,  0,  1
  0,  0,  1,  1,  0
]
删除边 1-3 后，图为
顶点列表 = 
1,3,2,5,4
邻接矩阵 =
[
  0,  0,  1,  1,  0
  0,  0,  1,  0,  0
  1,  1,  0,  1,  1
  1,  0,  1,  0,  1
  0,  0,  1,  1,  0
]
添加顶点 6 后，图为
顶点列表 = 
1,3,2,5,4,6
邻接矩阵 =
[
  0,  0,  1,  1,  0,  0
  0,  0,  1,  0,  0,  0
  1,  1,  0,  1,  1,  0
  1,  0,  1,  0,  1,  0
  0,  0,  1,  1,  0,  0
  0,  0,  0,  0,  0,  0
]
删除顶点 3 后，图为
顶点列表 = 
1,2,5,4,6
邻接矩阵 =
[
  0,  1,  1,  0,  0
  1,  0,  1,  1,  0
  1,  1,  0,  1,  0
  0,  1,  1,  0,  0
  0,  0,  0,  0,  0
]
```

 -----
 
 ```
 hello_algo.chapter_heap.heap.Test:
    Outcome: Passed
    Standard Output Messages:
    以下测试样例为大顶堆
堆顶元素为 5
堆元素数量为 0
堆是否为空 True
输入列表并建立小顶堆后
堆的数组表示：
1,2,3,4,5
堆的树状表示：
    /——— 3
——— 1
   |    /——— 5
    \——— 2
        \——— 4
        
-----

hello_algo.chapter_heap.my_heap.Test:
    Outcome: Passed
    Standard Output Messages:
    
输入列表并建堆后
堆顶元素为 9
元素 7 入堆后
堆的数组表示：
9,8,7,7,6,6,6,5,4,3,2,2,1
堆的树状表示：
        /——— 6
    /——— 7
   |   |    /——— 1
   |    \——— 6
   |        \——— 2
——— 9
   |        /——— 2
   |    /——— 6
   |   |    \——— 3
    \——— 8
       |    /——— 4
        \——— 7
            \——— 5
堆顶元素 9 出堆后
堆的数组表示：
8,7,7,6,6,6,5,4,3,2,2,1
堆的树状表示：
        /——— 5
    /——— 7
   |    \——— 6
   |        \——— 1
——— 8
   |        /——— 2
   |    /——— 6
   |   |    \——— 2
    \——— 7
       |    /——— 3
        \——— 6
            \——— 4
堆元素数量为 12
堆是否为空 False
```

---

File: codes/csharp/chapter_graph/graph_adjacency_list.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
+using NUnit.Framework;
```
Comment (krahets) [inline codes/csharp/chapter_graph/graph_adjacency_list.cs:1]: Please add the file header with your info.

File: codes/csharp/chapter_graph/graph_adjacency_matrix.cs [hunk 2]
```diff
@@ -0,0 +1,139 @@
+using hello_algo.include;
```
Comment (krahets) [inline codes/csharp/chapter_graph/graph_adjacency_matrix.cs:1]: Same as above.

File: codes/csharp/chapter_graph/graph_adjacency_list.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+namespace hello_algo.chapter_graph;
+/* 顶点类 */
```
Comment (krahets) [inline codes/csharp/chapter_graph/graph_adjacency_list.cs:4]: Add an empty line above

File: codes/csharp/chapter_graph/graph_adjacency_list.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        foreach (Vertex[] edge in edges)
+        {
+            AddVertex(edge[0]);
+            AddVertex(edge[1]);
+            AddEdge(edge[0], edge[1]);
+        }
+    }
+
+    /* 获取顶点数量 */
+    public int Size()
```
Comment (krahets) [inline codes/csharp/chapter_graph/graph_adjacency_list.cs:34]: size()

Same as others.

File: codes/csharp/chapter_graph/graph_adjacency_list.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            throw new InvalidOperationException();
+        // 在邻接表中删除顶点 vet 对应的链表（即 HashSet）
+        adjList.Remove(vet);
+        // 遍历其它顶点的链表（即 HashSet），删除所有包含 vet 的边
+        foreach (HashSet<Vertex> set in adjList.Values)
+        {
+            set.Remove(vet);
+        }
+    }
+    public void Print()
```
Comment (krahets) [inline codes/csharp/chapter_graph/graph_adjacency_list.cs:81]: Add an empty line above.

File: codes/csharp/chapter_graph/graph_adjacency_list.cs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+               v4 = new Vertex(4);
+        Vertex[][] edges = new Vertex[][] { new Vertex[] { v0, v1 }, new Vertex[] { v1, v2 }, new Vertex[] { v2, v3 }, new Vertex[] { v0, v3 }, new Vertex[] { v2, v4 }, new Vertex[] { v3, v4 } };
+        GraphAdjList graph = new GraphAdjList(edges);
+        Console.WriteLine("\n初始化后，图为");
+        graph.Print();
+
+        /* 添加边 */
+        // 顶点 1, 2 即 v0, v2
+        graph.AddEdge(v0, v2);
+        Console.WriteLine("添加边 1-2 后，图为");
```
Comment (krahets) [inline codes/csharp/chapter_graph/graph_adjacency_list.cs:113]: "\n添加边 1-2 后，图为"

Same as other strings for print.

File: codes/csharp/chapter_heap/my_heap.cs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        (maxHeap[i], maxHeap[p]) = (maxHeap[p], maxHeap[i]);
+    }
+    /* 打印堆（二叉树） */
+    public void Print()
+    {
+        PriorityQueue<int, int> queue = new PriorityQueue<int, int>(maxHeap.Select(x => (x, x)), Comparer<int>.Create((x, y) => y - x));
+        PrintUtil.PrintHeap(queue);
+    }
+}
+public class my_heap
```
Comment (krahets) [inline codes/csharp/chapter_heap/my_heap.cs:134]: Add an empty line above.

Same as others.

File: codes/csharp/include/PrintUtil.cs [hunk 8]
```diff
@@ -119,6 +119,29 @@ public static void showTrunks(Trunk? p)
                 Console.WriteLine(kv.ToString() + " -> " + map[kv]?.ToString());
             }
         }
-    }
 
+        public static void PrintHeap(PriorityQueue<int, int> heap)
+        {
+            Console.WriteLine("堆的数组表示：");
```
Comment (krahets) [inline codes/csharp/include/PrintUtil.cs:125]: Do not enter a new line here.

Write("堆的数组表示：");

File: docs/chapter_graph/graph_operations.md [hunk 9]
```diff
@@ -153,6 +153,94 @@ comments: true
...
+        // 参数 i, j 对应 vertices 元素索引
+        public void RemoveEdge(int i, int j)
+        {
+            // 索引越界与相等处理
+            if (i < 0 || j < 0 || i >= Size() || j >= Size() || i == j)
+                throw new IndexOutOfRangeException();
+            adjMat[i][j] = 0;
+            adjMat[j][i] = 0;
+        }
+    }
```
Comment (krahets) [inline docs/chapter_graph/graph_operations.md:244]: Remove the empty line.

File: docs/chapter_graph/graph_operations.md [hunk 10]
```diff
@@ -306,6 +394,83 @@ comments: true
...
+                throw new InvalidOperationException();
+            // 在邻接表中删除顶点 vet 对应的链表（即 HashSet）
+            adjList.Remove(vet);
+            // 遍历其它顶点的链表（即 HashSet），删除所有包含 vet 的边
+            foreach (HashSet<Vertex> set in adjList.Values)
+            {
+                set.Remove(vet);
+            }
+        }
+    }
```
Comment (krahets) [inline docs/chapter_graph/graph_operations.md:474]: Remove the empty line.

File: docs/chapter_heap/heap.md [hunk 11]
```diff
@@ -856,7 +976,18 @@ comments: true
 === "C#"
 
     ```csharp title="my_heap.cs"
-
+    /* 构造函数，根据输入列表建堆 */
+     public MaxHeap(IEnumerable<int> nums)
```
Comment (krahets) [inline docs/chapter_heap/heap.md:980]: Incorrect indentation

**Review by krahets — COMMENTED** [review]
Thanks for the PR! Please check the code carefully.

**Comment by krahets** [comment]
The outputs of `my_heap.cs` are not correct. Have you tested the codes and compared the results with Java's outputs?

It is a book for beginners, so let's check its correctness carefully.

**Comment by krahets** [comment]
Hi! The default branch is changed to from 'master' to 'main'.

From the local clone of the repository on a computer, run the following commands to update the name of the default branch.

```shell
git branch -m master main
git fetch origin
git branch -u origin/main main
git remote set-head origin -a
```

Optionally, run the following command to remove tracking references to the old branch name.

```shell
git remote prune origin
```

File: codes/csharp/chapter_heap/heap.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+        Console.WriteLine($"堆元素数量为 {size}");
+
+        /* 判断堆是否为空 */
+        bool isEmpty = maxHeap.Count == 0;
+
+        Console.WriteLine($"堆是否为空 {isEmpty}");
+
+        /* 输入列表并建堆 */
+        minHeap = new PriorityQueue<int, int>(new List<(int, int)> { (1, 1), (3, 3), (2, 2), (5, 5), (4, 4), });
```
Comment (krahets) [inline codes/csharp/chapter_heap/heap.cs:68]: Why double the variables?

Comment (ZJKung) [reply codes/csharp/chapter_heap/heap.cs:68]: In C# built in PriorityQueue there are two arguments,the first is element value and the second is it's priority.

File: codes/csharp/include/PrintUtil.cs [hunk 13]
```diff
@@ -1,9 +1,8 @@
...
- * File: PrintUtil.cs
- * Created Time: 2022-12-23
- * Author: haptear (haptear@hotmail.com)
- */
-
+* File: PrintUtil.cs
+* Created Time: 2022-12-23
+* Author: haptear (haptear@hotmail.com)
+*/
 namespace hello_algo.include;
```
Comment (krahets) [inline codes/csharp/include/PrintUtil.cs:6]: Add an empty line above

**Review by krahets — APPROVED** [review]
Thanks @ZJKung ! Now merge it.

**Review by krahets — APPROVED** [review]

**Pull request merged**